Add the ability to sync kube config files

Nokel81 opened the pull request:
- Will update when the files changes

To Dos:

- [x] Rework picking UI
- [x] ~Support proxy information?~
- [x] Add removal UI
- [x] Add label of the source file for filtering purposes
- [x] Support globs or watching folders?
- [x] Default watching of the default kubeconfig file

Signed-off-by: Sebastian Malton <sebastian@malton.name>

fixes: #2507 

UI
---
![Screen Shot 2021-04-23 at 9 35 37 AM](https://user-images.githubusercontent.com/8225332/115879171-5e67dc00-a417-11eb-9373-548a2a33d63d.png)

jakolehm commented on src/common/catalog-entity-registry.ts:
```diff
@@ -1,10 +1,14 @@
+    this.sources.set(id, computed(() => source));
+  }
+
+  @action addComputedSource(id: string, source: IComputedValue<CatalogEntity[]>) {
```
I think we should add computed support to `addSource` or rename `addSource` -> `addObservableSource` just to match naming.

jakolehm commented on src/main/catalog-sources/kubeconfig-sync.source.ts:
```diff
@@ -0,0 +1,223 @@
+
+type Disposer = () => void;
+
+export class KubeconfigSyncManager extends Singleton {
```
Maybe add tests for this class?

Nokel81 committed:
Add the ability to sync kube config files

- Will update when the files changes

- add KUBECONFIG_SYNC label

- fix rebase and change to addObservableSource

- move UI to user settings

- support shallow folder watching

- add some unit tests for the diff-er

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm commented on src/renderer/bootstrap.tsx:
```diff
@@ -90,9 +91,41 @@ export async function bootstrap(App: AppComponent) {
   </>, rootElem);
 }
 
+const defaultTheme = createMuiTheme({
```
What is this and why styling is done via "MuiTheme"?

Nokel81 replied:
This is how we change the defaults for material ui. This basically adds a new `<style>` to our DOM at the bottom of the `<head>` so it has the highest priority.

jakolehm replied:
I think we have done this differently in the past (through scss). @aleksfront @ixrock  wdyt about this?

jakolehm replied:
We didn't get any feedback on this... at least move this styling to a separate file and import it here (just to keep things cleaner).

Nokel81 replied:
okay

jakolehm commented on src/main/cluster-manager.ts:
```diff
@@ -59,17 +59,17 @@ export class ClusterManager extends Singleton {
   }
 
   @action protected updateCatalogSource(clusters: Cluster[]) {
```
Should we remove whole `updateCatalogSource`  because all local clusters in the catalog should come from "discovery"? We could migrate existing clusters by adding all kubeconfig files to watch list, right?

Nokel81 replied:
Yes I think so. Yes we could do that, though since we are already watching `ClusterStore.storedKubeConfigFolder` and `~/.kube/` we probably will get most of them anyway...

Nokel81 replied:
Will do is separate PR as discussed.

jakolehm commented on package.json:
```diff
@@ -171,6 +171,7 @@
     "@hapi/call": "^8.0.0",
     "@hapi/subtext": "^7.0.3",
     "@kubernetes/client-node": "^0.12.0",
+    "@material-ui/icons": "^4.11.2",
```
Should be in `devDependencies`.

Nokel81 replied:
Okay

jakolehm commented on src/common/catalog/catalog-entity-registry.ts:
```diff
@@ -1,10 +1,15 @@
 
 export class CatalogEntityRegistry {
-  protected sources = observable.map<string, IObservableArray<CatalogEntity>>([], { deep: true });
+  protected sources = observable.map<string, IComputedValue<CatalogEntity[]>>([], { deep: true });
```
Does `IComputedValue` cover both computed and observable?

Nokel81 replied:
Yes because an observable array can be trivially turned into a computed, see below

jakolehm commented on src/common/user-store.ts:
```diff
@@ -71,14 +76,32 @@ export class UserStore extends BaseStore<UserStoreModel> {
     // refresh new contexts
-    this.refreshNewContexts();
-    reaction(() => this.kubeConfigPath, this.refreshNewContexts);
+    await this.refreshNewContexts();
```
Just a note that we probably should remove `refreshNewContexts` because that information is not used anymore.

Nokel81 replied:
Will do in follow up PR

jakolehm commented on src/main/catalog-sources/index.ts:
```diff
@@ -0,0 +1 @@
+export { KubeconfigSyncManager } from "./kubeconfig-sync.source";
```
I think we don't need `.source` in the filename. Folder name is `catalog-sources`.

Nokel81 replied:
sounds good

jakolehm commented on src/renderer/components/app.scss:
```diff
@@ -55,7 +55,7 @@
 
 html {
-  font-size: 62.5%; // 1 rem == 10px
+  font-size: var(--font-size);
```
Out of scope for this PR?

Nokel81 replied:
I needed this to do the styling correctly for icons. This shouldn't break anything of ours since we don't use `rem`.

jakolehm replied:
Does not sound right, could you revert this and add more specific styling for icons?

Nokel81 replied:
Sure, do you know why we want the rem to be smaller?

Nachasic commented on src/common/__tests__/kube-helpers.test.ts:
```diff
@@ -72,55 +72,63 @@ describe("kube helpers", () => {
       describe("with valid kubeconfig", () => {
-        it("does not raise exceptions", () => {
-          expect(() => { validateKubeConfig(kc, "valid");}).not.toThrow();
+        it("does not return an error", () => {
```
Appreciate the phrasing.

Nachasic commented on package.json:
```diff
@@ -249,7 +250,7 @@
     "@types/html-webpack-plugin": "^3.2.3",
     "@types/http-proxy": "^1.17.5",
-    "@types/jest": "^25.2.3",
+    "@types/jest": "^26.0.22",
```
For some reason `yarn` keeps complaining about the integrity of jest types (from `make dev`):
```
yarn check --verify-tree --integrity
yarn check v1.22.10
error "@types/jest#pretty-format#@jest/types#@types/istanbul-reports" is wrong version: expected "^3.0.0", got "1.1.2"
error Found 1 errors.
```
I only was able to build this branch by updating dependency on `ts-jest` from `26.0.1` to `26.3.0`

Nachasic commented on src/common/kube-helpers.ts:
```diff
@@ -28,28 +28,41 @@ function readResolvedPathSync(filePath: string): string {
+}
+
+function checkRawUser(rawUser: any): boolean {
+  return rawUser?.name;
```
Personally, I'm in favor of explicit type conversion here.
```suggestion
  return Boolean(rawUser?.name);
```

Nachasic commented on src/common/kube-helpers.ts:
```diff
@@ -182,7 +207,7 @@ export function getNodeWarningConditions(node: V1Node) {
 /**
  * Checks if `config` has valid `Context`, `User`, `Cluster`, and `exec` fields (if present when required)
```
Perhaps there should be an annotation about this function returning exceptions instead of throwing them.

Nokel81 committed:
responding to review comments

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/common/__tests__/kube-helpers.test.ts:
```diff
@@ -72,55 +72,63 @@ describe("kube helpers", () => {
       describe("with invalid context object", () => {
-        it("it raises exception", () => {
-          expect(() => { validateKubeConfig(kc, "invalid");}).toThrow("No valid context object provided in kubeconfig for context 'invalid'");
+        it("it returns an error", () => {
```
While you're at it you could remove the extraneous "it" from the following test descriptions
```suggestion
        it("returns an error", () => {
```

jim-docker commented on src/common/catalog/catalog-entity-registry.ts:
```diff
@@ -1,10 +1,15 @@
+  protected sources = observable.map<string, IComputedValue<CatalogEntity[]>>([], { deep: true });
 
-  @action addSource(id: string, source: IObservableArray<CatalogEntity>) {
+  @action addObservableSource(id: string, source: IObservableArray<CatalogEntity>) {
+    this.sources.set(id, computed(() => source));
+  }
```
Not really relevant to this PR but I bring this up here because you are changing this code. I find the term source too vague. It's an "independent" or sourced list of entities, right? This makes it a little clearer, and we could do the same for the exposed API?
```suggestion
  @action addObservableSource(id: string, entities: IObservableArray<CatalogEntity>) {
    this.sources.set(id, computed(() => entities));
  }
```
or call it `addObservableEntitySource`?

Nokel81 replied:
So these are used `catalogEntityRegistry.addObservableSource(...)`. Do we need to have the word "Entity" twice, namely: `catalogEntityRegistry.addObservableEntitySource(...)` ?

jakolehm replied:
> Do we need to have the word "Entity" twice

No 🙂

jim-docker commented on src/common/catalog/catalog-entity-registry.ts:
```diff
@@ -1,10 +1,15 @@
+    this.sources.set(id, computed(() => source));
+  }
+
+  @action addComputedSource(id: string, source: IComputedValue<CatalogEntity[]>) {
     this.sources.set(id, source);
   }
```
```suggestion
  @action addComputedSource(id: string, entities: IComputedValue<CatalogEntity[]>) {
    this.sources.set(id, entities);
  }
```

jim-docker commented on src/common/catalog/catalog-entity-registry.ts:
```diff
@@ -13,7 +18,7 @@ export class CatalogEntityRegistry {
 
   @computed get items(): CatalogEntity[] {
-    return Array.from(this.sources.values()).flat();
+    return Array.from(iter.flatMap(this.sources.values(), source => source.get()));
```
```suggestion
    return Array.from(iter.flatMap(this.sources.values(), entity => entity.get()));
```

Nokel81 replied:
I don't think this specific name change is very good, because we are flat mapping over `sources.values()`. Also `sources` shouldn't really be called `entities` because it is a map between IDs and `IComputedValue<CatalogEntity[]>`.

jim-docker replied:
I guess conceptually it is a source but it is confusing for me because there are two different data types you are calling "source". You're not iterating over the `this.sources` map, but rather `this.sources.values()`, which gives you an iterable over just the entities, free of the IDs, right? I still like `entities` (not `entity`, my mistake) better here than `source`. Anyway, it's fine

Nokel81 replied:
Yes I am iterating over `this.sources.values()` but that is of the type `Iterable<CatalogEntities[]>` which is why I do a `iter.flatMap`

jim-docker commented on src/common/kube-helpers.ts:
```diff
@@ -181,8 +206,10 @@ export function getNodeWarningConditions(node: V1Node) {
 /**
  * Checks if `config` has valid `Context`, `User`, `Cluster`, and `exec` fields (if present when required)
+ *
+ * Note: This function returns an error instead of throwing it, returning `undefined` if the validation passes
```
Should you `try`/`catch` then in here, in case any of the called functions throw?

Nokel81 replied:
Probably

jim-docker commented on src/common/user-store.ts:
```diff
@@ -230,32 +243,44 @@ export class UserStore extends BaseStore<UserStoreModel> {
+    if (preferences.syncKubeconfigEntries) {
+      this.syncKubeconfigEntries.replace(preferences.syncKubeconfigEntries.map(({ filePath }) => filePath));
+    } else {
+      this.syncKubeconfigEntries.clear();
```
Don't we want it to always have ".kube" and `ClusterStore.storedKubeConfigFolder`?

Nokel81 replied:
Well if a user removes them we don't.

jakolehm replied:
I think we want to keep at least `ClusterStore.storedKubeConfigFolder` always. Otherwise "add cluster from c&p" will be seriously broken.

Nokel81 replied:
Okay, understood

jim-docker reviewed (commented):
about halfway through...

Nokel81 committed:
fix tests and add try/catch

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/common/user-store.ts:
```diff
@@ -71,14 +76,32 @@ export class UserStore extends BaseStore<UserStoreModel> {
+  syncKubeconfigEntries = observable.set([
+    path.join(os.homedir(), ".kube"),
+    ClusterStore.storedKubeConfigFolder,
+  ]);
```
Is this useful, since this set is cleared the first time `fromStore()` is called (assuming the store does not initially have `preferences.syncKubeconfigEntries`)?

jim-docker replied:
I stand corrected, at least when you start with an empty app data folder then these syncs are added. Another question though, should the `ClusterStore.storedKubeConfigFolder` even be visible to the user on the Preferences page? Shouldn't we always, and invisibly, sync this folder?

jim-docker commented on src/migrations/user-store/5.0.0-alpha.2.ts:
```diff
@@ -0,0 +1,18 @@
+
+    preferences.hiddenTableColumns = Object.entries(oldHiddenTableColumns);
+
+    store.get("preferences", preferences);
```
`store.set`?  (deja vu)

jakolehm replied:
This needs to be fixed.

Nokel81 replied:
Agreed, I am sorry that this slipped in. It almost certainly was the result of rebasing against the wrong branch target.

jakolehm commented on src/migrations/user-store/5.0.0-alpha.2.ts:
```diff
@@ -0,0 +1,18 @@
+import { migration } from "../migration-wrapper";
+
+export default migration({
+  version: "5.0.0-alpha.2",
```
We should bump this version, right?

Nokel81 replied:
This really shouldn't be here at all, will delete these two files.

Nokel81 committed:
always sync c&p folder, remove bad rebase

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix preferences

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/common/user-store.ts:
```diff
@@ -23,10 +22,12 @@ export interface UserStoreModel {
   filePath: string;
 }
 
+export interface KubeconfigSyncValue {}
```
for future use?

Nokel81 replied:
Yes, namely `httpProxy` settings

Nokel81 committed:
Fix settings saving and catalog view

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix unit tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix synced clusters not connectable

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/renderer/components/+preferences/kubeconfig-syncs.tsx:
```diff
@@ -0,0 +1,178 @@
+              {this.renderEntryIcon(entry)}
+            </Avatar>
+          </ListItemAvatar>
+          <ListItemText
+            primary={entry.filePath}
+            className="description"
+          />
```
I think the filePath text is too small
<img width="692" alt="Screen Shot 2021-04-28 at 7 25 15 PM" src="https://user-images.githubusercontent.com/40840436/116484704-a2b00d80-a857-11eb-9eaf-7509c39a0b7f.png">

Nokel81 replied:
I agree, will fix this.

jim-docker replied:
Could the folder icon be larger too, to fill more of the circular background?

jim-docker reviewed (commented):
- if I sync both a folder and a file inside the folder the clusters get added twice. Though this is  exactly what was asked for, it's not a great UX. Should we remove file syncs already covered by a folder sync?
- got this when adding a sync in Lens's app data:
<img width="692" alt="Screen Shot 2021-04-28 at 7 24 45 PM" src="https://user-images.githubusercontent.com/40840436/116485387-156db880-a859-11eb-845e-ffefc20ef0b4.png">
Of course I shouldn't be doing that...

Nokel81 committed:
change to non-complete shallow watching

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix sizing

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Catch readStream errors

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm commented on src/common/utils/iter.ts:
```diff
@@ -23,3 +23,67 @@ export function* take<T>(src: Iterable<T>, n: number): Iterable<T> {
     break outer;
   }
 }
+
```
Just a note: would make sense to add tests for these utils if they are used in many places.

Nokel81 replied:
I can do that in a separate PR, but good idea.

jakolehm commented on src/renderer/api/__tests__/catalog-entity-registry.test.ts:
```diff
@@ -43,27 +43,6 @@ describe("CatalogEntityRegistry", () => {
       expect(catalog.items.length).toEqual(2);
     });
 
-    it("ignores unknown items", () => {
```
Why remove? Makes no sense anymore?

Nokel81 replied:
Because I removed the merge logic, so the `renderer` just takes the whole blast from `main` and uses that.

jakolehm commented on src/main/catalog-sources/kubeconfig-sync.ts:
```diff
@@ -0,0 +1,249 @@
+
+          const entity = catalogEntityFromCluster(cluster);
+
+          entity.metadata.labels.KUBECONFIG_SYNC = filePath;
```
I think we should not add this label (unless we really have to). It adds too much noise to catalog.

Nokel81 committed:
don't clear UserStore on non-existant preference field, fix unlinking not removing items from source

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm reviewed (commented):
For unknown reason I did not get the default `~/.kube` folder synced when I started the app.

Nokel81 committed:
change label to file

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm reviewed (approved):
LGTM
